Search: Add Search module

nickdaugherty opened the pull request:
Adds the Search module, which offloads site search to WordPress.com, powered by Elasticsearch.

Includes Aggregation (Facets) support and corresponding Widget.

Is currently ‘headless’ - no admin UI exists so the module must be enabled manually.

kraftbj commented on modules/search/class.jetpack-search-widget-filters.php:
```diff
@@ -0,0 +1,125 @@
+
+		parent::__construct(
+			'jetpack-search-filters',
+			apply_filters( 'jetpack_widget_name', esc_html__( 'Search Filters (Jetpack)', 'jetpack' ) ),
```
Needs inline docs

nickdaugherty replied:
✅ Done

nickdaugherty replied:
✅ Done

kraftbj commented on modules/search/class.jetpack-search.php:
```diff
@@ -0,0 +1,1289 @@
+	 *
+	 * @module search
+	 */
+	public function setup() {
```
Should there be a check here to ensure the site is at the right plan level or confirm that an ES index exists? I assume the ES check is too expensive to run as often as this would run, but add an option to indicate index is pending or set that is checked on a regular basis?

gibrown replied:
FYI, an easy, lightweight way to verify would be to run a match_all query to the api and just get a single doc back. But ya you'd want to cache that or something.

nickdaugherty replied:
Yeah I've thought about that some - it's not _too_ much of a problem at this point because the module must be enabled manually and search will fall back to regular posts if the API returns an error, but I think it's definitely something we need to make more robust.

I wonder if we should implement this safeguard when we make a wider rollout of this (UI, auto index creation, etc). I'm ok with either way depending on how important we feel it is.

kraftbj commented on modules/search/class.jetpack-search.php:
```diff
@@ -0,0 +1,1289 @@
+
+		$this->do_search( $query );
+
+		if ( ! is_array( $this->search_result ) ) {
```
If I'm reading through this right, if the API fails for any reason -- no index, WP.com down, etc, this would have the site return traditional search results, correct?

nickdaugherty replied:
Yep! It will fall back to a regular search query when something goes wrong. Usually that will manifest itself as `$this->search_result` being a `WP_Error` from a bad HTTP request.

nickdaugherty replied:
Improved this behavior in 1ce891acbf5595e39fd247180aef768e9cd9eebb to handle more cases.

kraftbj commented on modules/search/class.jetpack-search.php:
```diff
@@ -0,0 +1,1289 @@
+								$name = $year;
+
+								// Is this year currently selected?
+								if ( ! empty( $query->get( 'year' ) ) && $query->get( 'year' ) === $year ) {
```
This isn't 5.2-5.4 compat. Need to set `$query->get( 'year' )` as a variable and then check if it is empty. This is one of the things failing the Travis tests.

nickdaugherty replied:
✅ Done

kraftbj commented on modules/search.php:
```diff
@@ -0,0 +1,14 @@
+ * Additional Search Queries: search
+ */
+
+require_once( __DIR__ . '/search/class.jetpack-search.php' );
```
Use `dirname( __FILE__ )` instead to make 5.2-friendly.

nickdaugherty replied:
✅ Done

MichaelArestad commented:
Do we need any admin or settings for this? Does it replace the WordPress search widget or add a new one like "Jetpack search" or something?

gibrown commented:
@MichaelArestad this has no ui at all. The only way to use it would be to write some code to enable it.

gibrown commented:
@jeherve built you an index for that blog_id

nickdaugherty commented on modules/search/class.jetpack-search-widget-filters.php:
```diff
@@ -0,0 +1,125 @@
+			$title = __( 'Filter By', 'jetpack' );
+		}
+
+		$title = apply_filters( 'widget_title', $title, $instance, $this->id_base );
```
✅ Done

nickdaugherty commented on modules/search/class.jetpack-search.php:
```diff
@@ -0,0 +1,1289 @@
+			'url'           => $service_url,
+		);
+
+		do_action( 'did_jetpack_search_query', $query );
```
✅ Done

nickdaugherty commented on modules/search/class.jetpack-search.php:
```diff
@@ -0,0 +1,1289 @@
+	 * @module search
+	 */
+	public function action__widgets_init() {
+		require_once( __DIR__ . '/class.jetpack-search-widget-filters.php' );
```
✅ Done

nickdaugherty commented:
Thanks for building that index @gibrown.

I _believe_ I've addressed all the feedback raised so far (thanks @jeherve and @kraftbj for the speedy turnaround!).

gibrown commented:
@nickdaugherty can you also include wpes-search-query-parser.php in this. It's completely generic and is perfect for helping folks to build their own custom search queries.

nickdaugherty commented:
@gibrown added the query parser lib in 804004de558fe955d59219a16aab1cfd71227c45.

jeherve commented:
> this has no ui at all. The only way to use it would be to write some code to enable it.

To clarify, there is currently a way to enable the Search module via the UI; the old module list, available at `https://yoursite.com/wp-admin/admin.php?page=jetpack_modules`, you can enable the module.

Not a lot happens when you do that, though: the module code doesn't run, as far as I can tell. The widget isn't registered.

I tried querying `/sites/%d/search` while authenticated with my Jetpack creds, but I only got a 404.

I'm not quite sure how to test this at this point :( If you have any pointers, let me know!

nickdaugherty commented:
@jeherve is there maybe something wrong with how this module is being registered? I ended up having to manually require the file on my VIP Go testing environment (I thought because of unrelated reasons - non-sandboxed requests switching out my active modules option).

On the modules admin page you linked to on my test site, there is a list item for Search, but the text is missing - it's just a checkbox with an empty area next to it, which makes me think this module is missing something in terms of getting JP to fully recognize it.

dereksmart commented:
> this module is missing something in terms of getting JP to fully recognize it.

@nickdaugherty everything is correct, it's just missing the build result of the `php:module-headings` [gulp task](https://github.com/Automattic/jetpack/blob/master/gulpfile.js#L566), which you can either run manually, or it should be generated by a fresh `yarn build`

nickdaugherty commented:
Makes sense, thanks @dereksmart!

@jeherve I'm not sure why it's not working for you - can you tell if it's getting past the checks in `Jetpack_Search::setup()`? https://github.com/Automattic/jetpack/pull/7208/files#diff-47deebe9b1fd7090caed8291af235f54R73

nickdaugherty commented:
Oh yes I do, fixed. 😄

jeherve commented:
@nickdaugherty That did it, thank you!

The widget doesn't display anything on the search result pages on my site though, because`$aggregations` is an empty array. Is there something I'd need to do so I can really test this?

nickdaugherty commented:
Desired Aggregations must be defined in code right now (but we should build a UI for it too), so this is a good test:

```
add_action( 'init', function() {
	if ( ! class_exists( 'Jetpack_Search' ) ) {
		return false;
	}

	Jetpack_Search::instance()->set_filters(array(
		'Content Type' => array(
			'type'     => 'post_type',
			'count'    => 10,
		),
		'Categories' => array(
			'type'     => 'taxonomy',
			'taxonomy' => 'category',
			'count'    => 10,
		),
		'Tags' => array(
			'type'     => 'taxonomy',
			'taxonomy' => 'post_tag',
			'count'    => 10,
		),
		'Year' => array(
			'type'     => 'date_histogram',
			'field'    => 'post_date',
			'interval' => 'year',
			'count'    => 10,
		),
		'Month' => array(
			'type'     => 'date_histogram',
			'field'    => 'post_date',
			'interval' => 'month',
			'count'    => 10,
		),
	) );
});
```

mjangda commented:
@jeherve @kraftbj @dereksmart Anything else we can do with this PR to get it ready in time for the 5.0 freeze? We're working on some developer-focused documentation that helps guide how to use this new module but let us know if there are other things we need to do as well.

dereksmart commented:
@mjangda We're prioritizing this for 5.0 👍  will be in touch if we need anything extra.

dereksmart commented on _inc/lib/jetpack-wpes-query-builder/jetpack-wpes-query-parser.php:
```diff
@@ -0,0 +1,671 @@
+				//from prefixes of "newspaper")
+				$must_q['bool']['should'][] = array( 'multi_match' => array(
+					'fields' => $this->all_fields,
+					'query' => $full_text,
```
`$full_text` doesn't seem to be defined?

dereksmart commented on modules/search/class.jetpack-search.php:
```diff
@@ -0,0 +1,1322 @@
+	 * @param WP_Query $query The original WP_Query to use for the parameters of our search
+	 */
+	public function do_search( WP_Query $query ) {
+		global $wpdb;
```
Is this used anywhere?

dereksmart commented:
Pushed a commit to hide the module from the legacy module list, and another commit to clean up some code style in one of the files

eliorivero commented on modules/search/class.jetpack-search.php:
```diff
@@ -0,0 +1,1314 @@
+				$aggregations_data[ $label ]['buckets'][] = array(
+					'url'        => add_query_arg( $url_params ),
+					'query_vars' => $query_vars,
+					'name'       => $name,
```
The variable $name might not have been defined. You probably want to initialize it to `''` and do something here if it's ever empty.

eliorivero commented on modules/search/class.jetpack-search.php:
```diff
@@ -0,0 +1,1314 @@
+			} // End foreach().
+		} // End foreach().
+
+		return $aggregations_data;
```
This var $aggregations_data might not have been defined but should be enough initializing it before the `foreach`.

nickdaugherty commented:
@jeherve That's odd - is it possible that the 2016 post that isn't showing up has a password or something? We add a `'perm' => 'readable'` to the posts query before surfacing it as a result, so wondering if that is somehow interfering.

https://github.com/Automattic/jetpack/blob/6148b5be7ac43f508e4fb98072a7a52d68a597e4/modules/search/class.jetpack-search.php#L198-L201

Regarding discrepancies between the admin search and frontend, likely this is because the admin search (just like core WP search) is matching everything that has the literal string `thing` in it, using a `LIKE '%thing%'` query...so posts with titles or content like `anything`, `something`, etc will match, where Elasticsearch recognizes that `anything` and `thing` are separate words / search terms.

jeherve commented:
>  is it possible that the 2016 post that isn't showing up has a password or something?

It is possible I suppose, although none of the posts I get for the "thing" search in wp-admin are password protected. In any case, maybe we shouldn't include it in the filter counts either?

> Regarding discrepancies between the admin search and frontend, likely this is because the admin search (just like core WP search) is matching everything that has the literal string thing in it

That makes sense 👍

nickdaugherty commented:
Those 'undefined index' warnings should be fixed in https://github.com/Automattic/jetpack/pull/7208/commits/32bd81eeb780889a9f7d980783794e2b9bd70543

nickdaugherty commented:
> It is possible I suppose, although none of the posts I get for the "thing" search in wp-admin are password protected. In any case, maybe we shouldn't include it in the filter counts either?

Ya those shouldn't be indexed at all as-is, I was just wondering if it could be a factor. 

So the missing result in your test is post id 2, "Sample Page". Looks like the query was only loading up `post` types, so I added `post_type => any` in https://github.com/Automattic/jetpack/pull/7208/commits/26f97813793a2eefa14e05a26a548b00600dd79e - can you please try again @jeherve?

eliorivero commented:
Added another commit checking for `buckets` index that was throwing notices in a couple of places.

jeherve commented:
> So the missing result in your test is post id 2, "Sample Page". Looks like the query was only loading up post types, so I added post_type => any in 26f9781 - can you please try again @jeherve?

That was fixed and it works now! Thank you!

dereksmart commented:
🐑
